### DISCUSSION TITLE
Moppy Microcontroller Control App

### DISCUSSION THREAD

**Discussion by Kroustalo** [SammyIAm/Moppy2#138] [Ideas]
Hello, I am looking to develop a version of the moppy GUI app to run on a microcontroller (AVR/ESP/Raspberry) instead of the Java app. 
I want to store the songs on the device, select them (using a lcd/button shield) and then, play on the same controller or send the data on a second microcontroller through serial to drive the moppy network. 

(Basically running the system without a computer) 

Is there any progress on this or I will start from scratch? 

Thanks in advance! :D

**Comment by samphonic** [comment]
Hi there, sorry it took me a bit to respond.

I know others have done this in the past, so I did a quick Google/GitHub search (there's probably more out there so it's worth doing a little searching yourself) and here's the first few things I came up with.

- @Ultrawipf [made one](https://github.com/Ultrawipf/midifloppy) way back in the day with the older version of Moppy.  Because of the age of the code I'm not sure it's suitable out of the box anymore, but could give you some inspiration.  He's got some videos on [his YouTube channel](https://www.youtube.com/channel/UCMjRDndeumfdaoEucOgm0xg).
- @dmadison has a[ library for converting the MIDI files to Moppy packets](https://github.com/dmadison/MIDItoMoppy), though again it's for the old Moppy so probably not directly useful.  If you're planning on using AVR, you might need to do this (see thoughts below).
- These German guys made [a really cool briefcase jukebox](https://www.youtube.com/watch?v=2RQE8DRJxWA) using a Raspberry Pi.  If you look close, you can see they're using Moppy2.0 code.  Can't find any more info about them though, so you'd have to track them down.


In general, the hardware platform will make the biggest difference in what you need to do:

**AVR** may struggle to store/decode the MIDI files *and* control the floppy drives.  You may need to pre-convert the MIDI files into something simpler for storage and reading.  Alternatively, using a separate microcontroller to control the floppy drives might let you dedicate an AVR to decoding MIDI.  I know some people[ have talked about](https://github.com/Sammy1Am/Moppy2/issues/62) an I2C implementation for linking a Moppy microcontrollers.
**ESP** should, I think, have enough power to decode MIDI files and also control the drives (though you might have some Timer issues maybe).  This is probably the route I would go if I were going to attempt this.  The Moppy2 code will already run on an ESP8266 or ESP32, so you'd still only need to add the controller/LCD code.
**Raspberry Pi**s are full-fledged computers, so you could just run the Java code on a tiny screen (though you'd need a mouse and keyboard I guess).  MoppyLib should have all the core control code separate from the GUI, so you could write your own LCD/button "ui" as a wrapper to MoppyLib.

tl;dr: You're probably best off starting from scratch rather than using other code directly, but there is some existing code around that should help guide you.

Let me know how it goes / if you run into any other questions!

**Reply by Kroustalo** [reply]
Alrighty, thank you for your information. You gave out the fundamentals to start.

I might go with the ESP (+?AVR control route). I'll check some stuff and propably update here or make my own repository (I have the whole summer in front of me :P) 

P.S. I need to check out the HDD melody that Gigawifp did...

**Reply by Ultrawipf** [reply]
My version is usb midi based and requires a computer (could be a raspberry pi) to send the notes to the USB midi device.

@Kroustalo
You could go with usb midi if you want it modular and standalone and use a raspberry pi with a headless midi player (could also be a simple python script with [mido](https://pypi.org/project/mido/)) to play songs to the usb midi port.
Might be easier to maintain than storing the songs on a microcontroller itself.

The versions in my repo still work fine and are actively used but you need an Arduino DUE and you may want to remove the LCD functions.
And for the updated version which can also control the enable lines 2 shift registers are required or you have to remove the `sendEnablePins` and related function calls depending on your setup.

**Reply by Kroustalo** [reply]
@Ultrawipf 
Thank you for taking notice 😃 Your implementation was one of my inspirations, and I really want to experiment with HDD tone as you did.

I would like to use a raspberry pi, but currently, I don't possess one (yet). I want to try my luck with the ESP/AVR Combo that I mentioned and then consider using it. 

~~Also I wanted to try something with OpenFFB, but I couldn't find a big enough motor. Guess I'll stick [racing ](https://youtu.be/4j63NB4-mww) with my T300 for a bit more...~~
